drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto err_out;
+    }
+    if (s->LocalAddressHandle != NULL)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = NULL;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+
+// DbgPrint("WskBind s is %p s->LocalAddressHandle is %p s->LocalAddressFile is %p status is %x TdiOpenAddressFile\n", s, s->LocalAddressHandle, s->LocalAddressFile, status);
+
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, TAG_NETIO);
+
+err_out:
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    PWSK_SOCKET_INTERNAL s = (PWSK_SOCKET_INTERNAL)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    FUNCTION_TRACE;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), TAG_NETIO);
+    if (nc == NULL)
+    {
+        goto err_out;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        goto err_out_free_nc;
+    }
+    nc->TargetConnectionInfo = TargetConnectionInfo;
+    nc->PeerAddrRet = NULL;
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        goto err_out_free_nc_and_tci;
+    }
+
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+
+    /* This will create a tdiIrp: */
+    status = TdiSendDatagram(&tdiIrp, s->LocalAddressFile, ((char *)BufferData) + Buffer->Offset,
+                        Buffer->Length, TargetConnectionInfo, NetioComplete, nc);
+
+    /* If allocating tdiIrp fails we get here.
+     * Call the IoCompletion of the application's Irp so this Irp
+     * gets freed.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        goto err_out_free_nc_and_tci_unmap;
+    }
+    return STATUS_PENDING;
+
+err_out_free_nc_and_tci_unmap:
+    /* TODO: implement freeing Buffer mmap */
+
+err_out_free_nc_and_tci:
+    ExFreePoolWithTag(TargetConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+
+err_out_free_nc:
+    ExFreePoolWithTag(nc, TAG_NETIO);
```

Comment:
Maybe consider (for this function, another similar above, and maybe others similar down below), to have a single "Quit:" or "Cleanup:" label, in the suggestion below I used `err_out_free`; instead of many different ones, and then do something like this:
```suggestion
err_out_free:
    if (TargetConnectionInfo)
        ExFreePoolWithTag(TargetConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
    if (nc)
        ExFreePoolWithTag(nc, TAG_NETIO);
```
This would require to initialize these pointers to NULL when they are first declared. (Then they get initialized later as usual.)

Reply:
Hi I would like to keep it that way. There are 1000+ ways to introduce new bugs and it works as it is ... I already did a lot of testing ... so unless you are strictly against it I would like to keep it like it is.